src/util/test/pagination-test.js

```diff
@@ -1,22 +1,129 @@
-import { pageNumberOptions } from '../pagination';
-
-describe('sidebar/util/pagination', () => {
-  describe('pageNumberOptions', () => {
-    [
-      { args: [1, 10, 5], expected: [1, 2, 3, 4, null, 10] },
-      { args: [3, 10, 5], expected: [1, 2, 3, 4, null, 10] },
-      { args: [6, 10, 5], expected: [1, null, 5, 6, 7, null, 10] },
-      { args: [9, 10, 5], expected: [1, null, 7, 8, 9, 10] },
-      { args: [2, 3, 5], expected: [1, 2, 3] },
-      { args: [3, 10, 7], expected: [1, 2, 3, 4, 5, 6, null, 10] },
-      { args: [1, 1, 5], expected: [] },
-    ].forEach(testCase => {
-      it('should produce expected available page numbers', () => {
-        assert.deepEqual(
-          pageNumberOptions(...testCase.args),
-          testCase.expected,
-        );
+import { paginationItems } from '../pagination';
+
+describe('paginationItems', () => {
+  [
+    // Only one page
+    {
+      current: 1,
+      total: 1,
+      expected: [],
+    },
+
+    // No pages elided.
+    {
+      current: 1,
+      total: 3,
+      expected: [1, 2, 3],
+    },
+
+    // Pages elided after current.
+    {
+      current: 1,
+      total: 4,
+      expected: [1, 2, null, 4],
+    },
+
+    // Pages elided before current.
+    {
+      current: 4,
+      total: 6,
+      expected: [1, null, 3, 4, 5, 6],
+    },
```

Comment:
This test is great, because it helps document the functionality, which is pretty complex.

However, it made me realize that, considering the elision in a 6-page pagination (`[1, null, 3, 4, 5, 6]`), I would have expected no elided pages for a 4-page pagination (`[1, 2, 3, 4]`) regardless of current page, but it seems we do elide (`[1, 2, null, 4]`).

Can you explain the algorithm to me as if I were a five-year-old 😅?

Reply:
In prose:

1. If there is one page, return an empty list
2. Calculate if there are elided pages before the current page
3. If no, create a list of all pages before the current page. If yes, create a list of the pages before the elided range, then `null`, then pages after the elided range.
4. Add the current page to the list
5. Calculate if there are elided pages after the current page
6. If no, append all pages after the current page. If yes, append a list of pages after the current page, before the elided range, then `null`, then pages after the elided range
7. Calculate the maximum number of items we might show, for all possible values of the `current` page
8. While the current number of items in the list is less than the maximum, expand the list by inserting an extra page number and "shrinking" the elided range. We preferentially take the page numbers from the elided range after the current page, but if that is empty, take them from the elided numbers before the current range.